### ISSUE TITLE
Spurious "useless lint attribute" warning

### ISSUE THREAD

**Issue by joshlf** [rust-lang-nursery/rust-clippy#1938]
Given the following code:

```rust
#[macro_use]
extern crate lazy_static;
```

`cargo build` produces the following warning:

```text
warning: unused `#[macro_use]` import
 --> src/lib.rs:1:1
  |
1 | #[macro_use]
  | ^^^^^^^^^^^^
  |
  = note: #[warn(unused_imports)] on by default
```

If we then add `#[allow(unused_imports)]`:

```rust
#[allow(unused_imports)]
#[macro_use]
extern crate lazy_static;
```

then `cargo build` produces no warnings. However, if we run `cargo clippy` on this code, we get a "useless lint attribute" warning:

```text
warning: useless lint attribute
 --> src/lib.rs:1:1
  |
1 | #[allow(unused_imports)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^ help: if you just forgot a `!`, use: `#![allow(unused_imports)]`
  |
  = note: #[warn(useless_attribute)] on by default
  = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#useless_attribute
```

**Comment by oli-obk** [comment]
Question: why do you want to ignore an unused `#[macro_use]`?

This is a known limitation of the lint (see https://rust-lang-nursery.github.io/rust-clippy/master/index.html#useless_attribute ), but as the limitation states, it's easier to remove the unused import than to add an `allow`.

**Comment by joshlf** [comment]
> Question: why do you want to ignore an unused `#[macro_use]`?

My crate exports macros that, internally, use `lazy_static`. Thus, when compiling the crate itself, nobody has called those macros, and thus `lazy_static` isn't actually used, so an unused import warning is generated.

> as the limitation states, it's easier to remove the unused import than to add an `allow`.

In my case, I actually need the import.

**Comment by oli-obk** [comment]
That sounds more like a rustc bug to me. Have you reported this upstream?

**Comment by joshlf** [comment]
I don't think it's a rustc bug. Maybe I'm missing something though. [Here's the full file](https://gist.github.com/joshlf/87783078dad1d888a321db50da5d4097) if you're curious.

**Comment by oli-obk** [comment]
Ah... your macro is unusable within crates that don't have 

```rust
#[macro_use]
extern crate lazy_static;
```

So I suggest to move the `#[macro_use] extern crate lazy_static;` at the beginning of your macro: https://gist.github.com/joshlf/87783078dad1d888a321db50da5d4097#file-lib-rs-L484

The trick mentioned in https://gist.github.com/joshlf/87783078dad1d888a321db50da5d4097#file-lib-rs-L474 kinda works but is hacky imo. The solution of moving the crate import into the macro itself is also used by `serde` and others, so I consider that solution idiomatic.

**Comment by joshlf** [comment]
I'm not sure that it would work. My concern is that since `#[macro_use]` is only allowed on `extern crate` declarations at the crate root, it would bar the macro from being used anywhere other than a crate's root. I'll have to try it out to see, though.

**Comment by joshlf** [comment]
Oh, and more problematically, putting `#[macro_use] extern crate lazy_static;` in the macro requires a `lazy_static` dependency in the consuming crate's `Cargo.toml`, which I definitely want to avoid.

**Comment by nathanaeljones** [comment]
This is a major pain with `#![feature(alloc_system)]` [see](https://doc.rust-lang.org/1.17.0/unstable-book/alloc-system.html)

rustc gives a spurious unused crate error for `extern crate alloc_system`.

When we disable that lint, `#[allow(unused_extern_crates)]`, clippy complains (incorrectly) about *that* attribute.

Which brings us to this mess:

```
#![feature(alloc_system)]
#[cfg_attr(feature = "cargo-clippy", allow(useless_attribute))]
#[allow(unused_extern_crates)]
extern crate alloc_system;
```

**Comment by llogiq** [comment]
The `unused_extern_crates` lint was made to warn by default recently, but this lead to many false positives. The newest (or next?) nightly should have a PR that disables the lint again until the false positives are sorted out.

**Comment by oli-obk** [comment]
> Oh, and more problematically, putting #[macro_use] extern crate lazy_static; in the macro requires a lazy_static dependency in the consuming crate's Cargo.toml, which I definitely want to avoid.

You have to do that anyway as far as I can see.

> When we disable that lint, #[allow(unused_extern_crates)], clippy complains (incorrectly) about that attribute.

That has been fixed in rustc

**Comment by joshlf** [comment]
>> Oh, and more problematically, putting #[macro_use] extern crate lazy_static; in the macro requires a lazy_static dependency in the consuming crate's Cargo.toml, which I definitely want to avoid.
> 
> You have to do that anyway as far as I can see.

Nah, I'm able to do:

```rust
#[doc(hidden)]
pub use lazy_static::*;
```

>> When we disable that lint, #[allow(unused_extern_crates)], clippy complains (incorrectly) about that attribute.
> 
> That has been fixed in rustc

Good to know!

**Comment by oli-obk** [comment]
In that case the lint `unused_imports` won't be triggered, right?

**Comment by brandur** [comment]
Another example of a false positive case: I use `serde_json` in my project, but I only need the macros it provides in my test suite. If I annotate it with `#[macro_use]`, Rustc correctly complains that I'm not using those macros (outside of tests).

The solution is to annotate the `extern crate` with `#[allow(unused_imports)]`, but unfortunately, this leads to Clippy complaining.

Whether you consider this a Rustc problem or not, it's pretty unambiguous that the ["why this is bad" description is wrong isn't it](https://rust-lang-nursery.github.io/rust-clippy/v0.0.187/index.html#useless_attribute)?

> Lint attributes have no effect on crate imports. Most likely a ! was forgotten

Having an `allow` there provably fixes a compiler warning, so it certainly has an effect.

**Comment by sinkuu** [comment]
BTW you can simply annotate it with `#[cfg_attr(test, macro_use)]`.

**Comment by brandur** [comment]
@sinkuu Thanks! I'd basically solved it with:

``` rust
#[cfg_attr(feature = "cargo-clippy", allow(useless_attribute))]
#[allow(unused_imports)]
#[macro_use]
extern create serde_json;
```

But that seems better.

**Comment by dtolnay** [comment]
As reported in #1995 and #2607, this false positive is triggered when re-exporting procedural macros.

```rust
#[allow(unused_imports)]
#[macro_use]
extern crate repro_derive;

pub use repro_derive::*;
```

**Comment by flip1995** [comment]
@gnzlbg `#[allow(unused_imports)]` is whitelisted for `use` items. The `unused_import` (without `s`) lint doesn't exist, see [rustc book](https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unused-imports).
